requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,192 @@
+
+To support the most common workflow of drafting dashboards/etc. in dev/staging and then "promoting" to production, entities such as dashboards would be environment-specific by default, but could be rolled out to the whole project with two clicks. When rolled out, the entity's dependencies would be made project-wide as well (e.g. for a cohort, all actions and other cohorts used in the definition would be instantly made project-wide as well). We'd introduce checks to make sure you can't make an entity environment-specific if it's currently being used project-wide. Sharing in this way would involve no copying though, so there would be no effort needed to sync things.
+
+To support large customers with access control needs, we'd need to add environment-specific project membership to the existing project access control system.
+
+To support the common feature flags workflow of rollout conditions being different between dev and production, we'd make flag rollout groups always environment-specific.
+
+To ensure straightforward yet complete separation of data at the ingestion level (incl. persons), *as well as* steady querying performance, each environment would have its own `team_id`. Due to this requirement, environments would have to be predefined by users and not created at ingestion time, thus each environment would also have its own SDK key (aka "project API key" currently). That seems like a safe choice in terms of the UX.
+
+By default, a project would only have one environment implicitly named "Production". A "New environment" button would allow adding a new one, with the environments feature explained in the modal. The current environment would then be selectable from the project selector, roughly looking like this (courtesy of Cory):
+
+Extra internal requirement: Because Team Pipeline plans to implement multiple SDK keys per project, it would be simplest to move SDK keys to their own model, rather than moving to the environment model.
+
+<img width="939" alt="image" src="https://github.com/PostHog/meta/assets/4550621/ec966c6b-0a1c-4cb0-a593-0234f4807ab1">
+
+### Downsides
+
+Of course, nothing comes for free: adding evironments has its own downsides. Here, it's the amount of effort required – this feature would touch every part of the app, plus ingestion. It would realistically be a few person-sprints of work (as always, it's hard to estimate accurately).
+
+### Pricing
+
+Having multiple environments should start out as a Teams/Enterprise feature.
+
+### Rough scope of work
+
+1. Backend: We add an `Environment` model (`posthog_environment` table, columns int `id`, `team_id`, `data_id` `name`) and hook it up to a new project-scoped viewset (`/api/projects/:id/environments/`). Every time a new environment is created, we determine its `id` by incrementing the counter of the `posthog_team.id` column and using that value as the environment's `data_id`.
```

Comment:
I don't want to add a lot of noise at this stage _but_ did you consider the idea of re-using projects for this?

Core idea:
1. Projects can be created as sub-projects of an existing one
2. The additional property `parent_team_id` is added to each Team
3. When you select an "environment" what you are actually doing is selecting a different project but the UI shows it as a sub project
4. Loading of dashboards/insights etc. use the `parent_team_id` instead with only the data specific endpoints going to the sub-project

The potential benefits would be:
1. API routes remain the same as before - everything is project scoped
2. UI routes work as before which is super nice (we don't need `current_environment` or eventually some way to include it in the URL)
3. No need to re-implement permissions - these would apply as before but just with a UI makeover to make it clear what has changed
4. You could import existing projects easily into a parent project giving us a simple migration path for existing users who want to group their staging / prod projects

This might need a little more fleshing out of course. Just wanted to propose it as it feels like a way faster way of getting to the same result without introducing a bunch of extra models.

Reply:
The main issues I see in this approach are
1. Project specific things like Insights with promotion (promoting a dashboard from staging to production). This would have to be done on the parent project still with some additional property or _in_ the sub project but then you have the copying issue
2. Shared ingestion-y type things such as Event/Property definitions - same could apply as in the proposed solution - if a `parent_team_id` is set then we process in that team's definitions instead (or both)

Reply:
Made a PR here to flesh out how I think this differs implementation-wise https://github.com/PostHog/meta/pull/185